File: app/src/main/java/ru/iandreyshev/parserrss/presentation/presenter/FeedPresenter.java
package ru.iandreyshev.parserrss.presentation.presenter;

import android.support.annotation.NonNull;

import ru.iandreyshev.parserrss.R;
import ru.iandreyshev.parserrss.app.App;
import ru.iandreyshev.parserrss.models.async.DeleteRssFromDbTask;
import ru.iandreyshev.parserrss.models.async.GetAllRssFromDbTask;
import ru.iandreyshev.parserrss.models.async.InsertNewRssTask;
import ru.iandreyshev.parserrss.models.filters.FilterByDate;
import ru.iandreyshev.parserrss.models.rss.ViewRss;
import ru.iandreyshev.parserrss.models.web.IHttpRequestResult;
import ru.iandreyshev.parserrss.presentation.view.IFeedView;

import com.arellomobile.mvp.InjectViewState;
import com.arellomobile.mvp.MvpPresenter;

import java.util.List;

@InjectViewState
public final class FeedPresenter extends MvpPresenter<IFeedView> {
    private long mProgressBarUserCount;

    public void onInsertRss(final String url) {
        InsertNewRssTask.execute(new InsertRssFromNetListener(), url, FilterByDate.Companion.newInstance());
    }

    public void onDeleteRss(final ViewRss rss) {
        DeleteRssFromDbTask.execute(new DeletingRssListener(), rss);
        getViewState().removeRss(rss);
    }

    public void openArticle(long articleId) {
        getViewState().openArticle(articleId);
    }

    public void openRssInfo(@NonNull final ViewRss rss) {
        getViewState().openRssInfo(rss);
    }

    @Override
    protected void onFirstViewAttach() {
        super.onFirstViewAttach();
        GetAllRssFromDbTask.execute(new LoadFromDatabaseListener(), FilterByDate.Companion.newInstance());
    }

    private void startProgressBar(boolean isStart) {
        mProgressBarUserCount += (isStart) ? 1 : -1;
        getViewState().startProgressBar(mProgressBarUserCount > 0);
    }

    private class DeletingRssListener implements DeleteRssFromDbTask.IEventListener {
        @Override
        public void onFail(ViewRss rss) {
            getViewState().showShortToast(App.getStr(R.string.toast_error_deleting_from_db));
        }

        @Override
        public void onPreExecute() {
        }

        @Override
        public void onPostExecute(final ViewRss result) {
        }
    }

    private class InsertRssFromNetListener implements InsertNewRssTask.IEventListener {
        @Override
        public void onInvalidUrl() {
            getViewState().showShortToast(App.getStr(R.string.toast_invalid_url));
        }

        @Override
        public void onParserError() {
            getViewState().showShortToast(App.getStr(R.string.toast_invalid_rss_format));
        }

        @Override
        public void onNetError(final IHttpRequestResult requestResult) {
            switch (requestResult.getState()) {
                case BadConnection:
                    getViewState().showShortToast(App.getStr(R.string.toast_bad_connection));
                    break;

                case PermissionDenied:
                    getViewState().showShortToast(App.getStr(R.string.toast_internet_permission_denied));
                    break;
            }
        }

        @Override
        public void onRssAlreadyExist() {
            getViewState().showShortToast(App.getStr(R.string.toast_rss_already_exist));
        }

        @Override
        public void onDatabaseError() {
            getViewState().showShortToast(App.getStr(R.string.toast_error_saving_to_db));
        }

        @Override
        public void onSuccess(final ViewRss rss) {
            getViewState().insertRss(rss);
        }

        @Override
        public void onPreExecute() {
            startProgressBar(true);
        }

        @Override
        public void onPostExecute(final ViewRss result) {
            startProgressBar(false);
        }
    }

    private class LoadFromDatabaseListener implements GetAllRssFromDbTask.IEventListener {
        @Override
        public void onLoadError() {
            getViewState().showShortToast(App.getStr(R.string.toast_error_loading_from_db));
        }

        @Override
        public void onSuccess(final List<ViewRss> rssFromDb) {
            for (final ViewRss rss : rssFromDb) {
                getViewState().insertRss(rss);
            }
        }

        @Override
        public void onPreExecute() {
            startProgressBar(true);
        }

        @Override
        public void onPostExecute(List<ViewRss> result) {
            startProgressBar(false);
        }
    }
}

File: app/src/main/kotlin/ru/iandreyshev/parserrss/presentation/presenter/FeedPresenter.kt
package ru.iandreyshev.parserrss.presentation.presenter

import ru.iandreyshev.parserrss.models.viewModels.ViewRss
import ru.iandreyshev.parserrss.presentation.view.IFeedView

import com.arellomobile.mvp.InjectViewState
import com.arellomobile.mvp.MvpPresenter
import ru.iandreyshev.parserrss.R
import ru.iandreyshev.parserrss.factory.useCase.UseCaseFactory
import ru.iandreyshev.parserrss.models.interactor.FeedInteractor
import ru.iandreyshev.parserrss.models.counter.ProcessCounter
import ru.iandreyshev.parserrss.models.useCase.feed.OpenArticleUseCase
import ru.iandreyshev.parserrss.models.useCase.feed.DeleteRssUseCase
import ru.iandreyshev.parserrss.models.useCase.feed.InsertRssUseCase
import ru.iandreyshev.parserrss.models.useCase.feed.LoadAllRssUseCase
import ru.iandreyshev.parserrss.models.useCase.feed.OpenRssInfoUseCase
import ru.iandreyshev.parserrss.models.web.HttpRequestHandler
import ru.iandreyshev.parserrss.models.web.IHttpRequestResult
import ru.iandreyshev.parserrss.presentation.presenter.extention.addRssDialog
import ru.iandreyshev.parserrss.presentation.presenter.extention.toast
import ru.iandreyshev.parserrss.presentation.presenter.extention.uiThread
import ru.iandreyshev.parserrss.ui.fragment.InternetPermissionDialog
import ru.iandreyshev.parserrss.ui.fragment.RssInfoDialog

@InjectViewState
class FeedPresenter : MvpPresenter<IFeedView>() {

    private val mProcessCounter = ProcessCounter(this::onChangeProcessCount)
    private val mRssCounter = ProcessCounter(this::onChangeRssCount)

    val interactor = FeedInteractor(UseCaseFactory(), UseCaseListener())

    private inner class UseCaseListener : LoadAllRssUseCase.IListener,
            DeleteRssUseCase.IListener,
            OpenArticleUseCase.IListener,
            InsertRssUseCase.IListener,
            OpenRssInfoUseCase.IListener {

        override fun loadRss(rss: ViewRss) = uiThread {
            viewState.insertRss(rss)
            mRssCounter.add()
        }

        override fun insertNewRss(rss: ViewRss, isFull: Boolean) = uiThread {
            viewState.insertRss(rss)
            viewState.openRssPage(rss.id)
            mRssCounter.add()
            onChangeCapacityStatus(isFull)
        }

        override fun rssCountIsMax() = uiThread {
            onChangeCapacityStatus(true)
        }

        override fun removeRss(id: Long) = uiThread {
            viewState.removeRss(id)
            mRssCounter.remove()
            onChangeCapacityStatus(false)
        }

        override fun openArticle(articleId: Long) = uiThread {
            viewState.openArticle(articleId)
        }

        override fun rssAlreadyExist() = uiThread {
            toast(R.string.toast_rss_already_exist)
        }

        override fun connectionError(requestResult: IHttpRequestResult) = uiThread {
            when (requestResult.state) {
                HttpRequestHandler.State.BAD_URL -> {
                    toast(R.string.toast_invalid_url)
                    viewState.openDialog(addRssDialog(requestResult.urlString))
                }
                HttpRequestHandler.State.PERMISSION_DENIED -> {
                    toast(R.string.toast_internet_permission_denied)
                    viewState.openDialog(InternetPermissionDialog())
                }
                else -> {
                    toast(R.string.toast_bad_connection)
                    viewState.openDialog(addRssDialog(requestResult.urlString))
                }
            }
        }

        override fun invalidRssFormat(url: String) = uiThread {
            toast(R.string.toast_invalid_rss_format)
            viewState.openDialog(addRssDialog(url))
        }

        override fun updateCapacityAfterLoad(isFull: Boolean) = uiThread {
            onChangeCapacityStatus(isFull)
        }

        override fun removingRssFailed() = uiThread {
            toast(R.string.feed_deleting_error)
        }

        override fun urlToAddRssIsEmpty() = uiThread {
            toast(R.string.feed_url_to_add_is_empty)
        }

        override fun openRssInfo(rss: ViewRss?) = uiThread {
            if (rss != null) {
                viewState.openDialog(RssInfoDialog.newInstance(rss))
            } else {
                toast(R.string.toast_open_info_fail)
            }
        }

        override fun processStart() = uiThread {
            mProcessCounter.add()
        }

        override fun processEnd() = uiThread {
            mProcessCounter.remove()
        }
    }

    private fun onChangeProcessCount(newCount: Int) {
        viewState.startProgressBar(newCount > 0)
    }

    private fun onChangeRssCount(newCount: Int) {
        viewState.setToolbarScrollable(newCount > 0)
        viewState.openEmptyContentMessage(newCount <= 0)
        viewState.openToolbarTitle(newCount <= 0)

        viewState.enableDeleteButton(newCount > 0)
        viewState.enableInfoButton(newCount > 0)
    }

    private fun onChangeCapacityStatus(isFull: Boolean) {
        viewState.enableAddButton(!isFull)
        if (isFull) {
            toast(R.string.feed_is_full)
        }
    }
}
